pkg/promotion/runner/builtin/yaml_merger.go

```diff
@@ -0,0 +1,131 @@
+package builtin
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	securejoin "github.com/cyphar/filepath-securejoin"
+	"github.com/xeipuuv/gojsonschema"
+
+	kargoapi "github.com/akuity/kargo/api/v1alpha1"
+	"github.com/akuity/kargo/pkg/promotion"
+	"github.com/akuity/kargo/pkg/x/promotion/runner/builtin"
+	"github.com/akuity/kargo/pkg/yaml"
+)
+
+const stepKindYAMLMerge = "yaml-merge"
+
+func init() {
+	promotion.RegisterStepRunner(
+		stepKindYAMLMerge,
+		promotion.StepRunnerRegistration{Factory: newYAMLMerger},
+	)
+}
+
+// yamlMerger is an implementation of the promotion.StepRunner interface that
+// merges multiple YAML files.
+type yamlMerger struct {
+	schemaLoader gojsonschema.JSONLoader
+}
+
+// newYAMLMerger returns an implementation of the promotion.StepRunner interface
+// that merges multiple YAML files.
+func newYAMLMerger(promotion.StepRunnerCapabilities) promotion.StepRunner {
+	return &yamlMerger{schemaLoader: getConfigSchemaLoader(stepKindYAMLMerge)}
+}
+
+// Run implements the promotion.StepRunner interface.
+func (y *yamlMerger) Run(
+	ctx context.Context,
+	stepCtx *promotion.StepContext,
+) (promotion.StepResult, error) {
+	cfg, err := y.convert(stepCtx.Config)
+	if err != nil {
+		return promotion.StepResult{
+			Status: kargoapi.PromotionStepStatusFailed,
+		}, &promotion.TerminalError{Err: err}
+	}
+	return y.run(ctx, stepCtx, cfg)
+}
+
+// convert validates yamlMerger configuration against a JSON schema and
+// converts it into a builtin.YAMLMergeConfig struct.
+func (y *yamlMerger) convert(cfg promotion.Config) (builtin.YAMLMergeConfig, error) {
+	return validateAndConvert[builtin.YAMLMergeConfig](y.schemaLoader, cfg, stepKindYAMLMerge)
+}
+
+func (y *yamlMerger) run(
+	_ context.Context,
+	stepCtx *promotion.StepContext,
+	cfg builtin.YAMLMergeConfig,
+) (promotion.StepResult, error) {
+
+	// Secure join the input paths to prevent path traversal attacks.
+	filePaths := []string{}
+	for _, path := range cfg.InFiles {
+		inFile, err := securejoin.SecureJoin(stepCtx.WorkDir, path)
+		if err != nil {
+			return promotion.StepResult{Status: kargoapi.PromotionStepStatusErrored},
+				fmt.Errorf("could not secure join input file %q: %w", path, err)
+		}
+
+		// only add existing files
+		_, err = os.Stat(inFile)
+		if err != nil {
+			if cfg.IgnoreMissingFiles {
+				continue
+			}
+			return promotion.StepResult{Status: kargoapi.PromotionStepStatusErrored},
+				fmt.Errorf("error accessing file %s: %w", inFile, err)
+
+		}
+		filePaths = append(filePaths, inFile)
+
+	}
```

Comment:
if all files are missing and `IgnoreMissingFiles` is true, this will pass an empty slice to `yaml.MergeFiles` which will fail